src/s3_fetchers.rs

```diff
@@ -1,84 +1,102 @@
             .collect(),
-    };
-
-    Ok(crate::types::ListObjectResponse {
-        continuation_token,
-        folder_names,
     })
 }
 
-/// By the given block height (`key`) gets the objects:
+/// By the given block height (`block_height_prefix`) gets the objects:
 /// - block.json
 /// - shard_N.json
 /// Reads the content of the objects and parses it to JSON.
 /// Returns the result in a single JSON
-pub(crate) async fn get_object(
-    client: &Client,
-    bucket: &str,
-    key: &str,
-    tracked_shards: &Vec<u8>,
-) -> Result<serde_json::Value, Error> {
-    let mut main_json = serde_json::json!({});
-    let block_json = {
-        let response = client
-            .get_object()
-            .bucket(bucket)
-            .key(format!("{}block.json", key))
-            .send()
-            .await?;
+pub(crate) async fn fetch_streamer_message(
+    s3_client: &Client,
+    s3_bucket_name: &str,
+    block_height_prefix: &str,
+) -> anyhow::Result<near_indexer_primitives::StreamerMessage> {
+    let block_json: serde_json::Value = {
+        let response = loop {
+            if let Ok(response) = s3_client
+                .get_object()
+                .bucket(s3_bucket_name)
+                .key(format!("{}block.json", block_height_prefix))
+                .send()
+                .await
+            {
+                break response;
+            }
+        };
 
         let body_bytes = response.body.collect().await.unwrap().into_bytes();
 
         serde_json::from_slice(body_bytes.as_ref()).unwrap()
     };
 
-    main_json["block"] = block_json;
+    let shards_num: u64 = block_json["header"]["chunks_included"].as_u64().unwrap();
+
+    let mut main_json = serde_json::json!({
+        "block": block_json,
+    });
 
     let mut shards: Vec<serde_json::Value> = vec![];
-    // TODO: undefined bahaviour in case if we track N > 1 shards
-    // but there is only N-1 shards, we need to handle it. Probably,
-    // by checking the `block.chunks` amount
-    for shard_id in tracked_shards {
-        let response = client
-            .get_object()
-            .bucket(bucket)
-            .key(format!("{}shard_{}.json", key, shard_id))
-            .send()
-            .await?;
 
-        let body_bytes = response.body.collect().await.unwrap().into_bytes();
+    let mut shards_futures: futures::stream::FuturesOrdered<_> = (0..shards_num)
+        .collect::<Vec<u64>>()
+        .into_iter()
+        .map(|shard_id| {
+            fetch_shard_or_retry(s3_client, s3_bucket_name, block_height_prefix, shard_id)
+        })
+        .collect();
 
-        shards.push(serde_json::from_slice(body_bytes.as_ref()).unwrap());
+    while let Some(shard) = shards_futures.next().await {
+        shards.push(shard.unwrap());
     }
 
     main_json["shards"] = serde_json::Value::Array(shards);
 
-    Ok(main_json)
+    Ok(serde_json::from_value::<near_indexer_primitives::StreamerMessage>(main_json).unwrap())
```

Comment:
... and then deserialize these into `IndexerShard`, and manually craft `StreamerMessage` from the `IndexerBlock` and `Vec<IndexerShard>`.

Reply:
I would suggest moving this into a separate issue until the `near-indexer-primitives` release

Reply:
There is nothing to wait for. I suggest we just use the underlying types we use anyway instead of messing with `serde_json::Value` and crafting `main_json`.